libc/stdio_file.c

```diff
@@ -0,0 +1,785 @@
+{ /* empty */
+}
+
+
+TEST(line, getline_basic)
+{
+	char *line = NULL;
+	size_t len = 1;
+
+	/* read using getline */
+	fd = fopen(stdpath, "r");
+	{
+		/* gerline with null buffer and misleading size */
+		TEST_ASSERT_EQUAL_INT(6, getline(&line, &len, fd));
+		TEST_ASSERT_EQUAL_STRING("line1\n", line);
+		/* new buffer shall be allocated of size at least strlen+1 */
+		TEST_ASSERT_GREATER_OR_EQUAL_INT(7, len);
+
+		/* getline with to small buffer */
+		TEST_ASSERT_EQUAL_INT(10, getline(&line, &len, fd));
+		TEST_ASSERT_EQUAL_STRING("lineline2\n", line);
+		/* buffer shall be reallocated of size at least strlen+1 */
+		TEST_ASSERT_GREATER_OR_EQUAL_INT(11, len);
+
+		/* getline with adequate buffer and character */
+		TEST_ASSERT_EQUAL_INT(6, getline(&line, &len, fd));
+		TEST_ASSERT_EQUAL_STRING("line3\n", line);
+		/* buffer shall not be reallocated, and stay at size at least as big as previusly*/
+		TEST_ASSERT_GREATER_OR_EQUAL_INT(11, len);
+
+		/* getline with adequate buffer, but only newline is read */
+		TEST_ASSERT_EQUAL_INT(1, getline(&line, &len, fd));
+		TEST_ASSERT_EQUAL_STRING("\n", line);
+		/* buffer shall not be reallocated, and stay at size at least as big as previusly*/
+		TEST_ASSERT_GREATER_OR_EQUAL_INT(11, len);
+
+		/* getline reading EOF */
+		TEST_ASSERT_EQUAL_INT(-1, getline(&line, &len, fd));
+		/* buffer shall not change from previous call */
+		TEST_ASSERT_EQUAL_STRING("\n", line);
+		/* buffer shall not be reallocated, and stay at size at least as big as previusly*/
+		TEST_ASSERT_GREATER_OR_EQUAL_INT(11, len);
+
+		free(line);
+	}
+	fclose(fd);
+}
+
+
+TEST(line, getline_wronly)
+{
+	char *line = NULL;
+	size_t len = 0;
+
+	/* read using getline from write-only file */
+	fd = fopen(stdpath, "a");
+	{
+		rewind(fd);
+		TEST_ASSERT_EQUAL_INT(-1, getline(&line, &len, fd));
+		TEST_ASSERT_EQUAL_INT(EBADF, errno);
+		TEST_ASSERT_NULL(line);
+	}
+	fclose(fd);
+}
+
+
+TEST(line, getline_allocated)
+{
+	/* If *lineptr is a null pointer or if the object pointed to by *lineptr is of insufficient size, 
+	an object shall be allocated as if by malloc() or the object shall be reallocated as if by realloc(), 
+	respectively, such that the object is large enough to hold the characters to be written to it, including the terminating NUL, 
+	and *n shall be set to the new size */
+	/* <posix incompliant> - value pointed to by *len argument should not change but is changed and equals strlen(line) */
+	//TEST_IGNORE();
+
+	char *line = NULL;
+	size_t len = 50; /* allocated memory exceeds one demanded for a line that to be read */
+
+	fd = fopen(stdpath, "r");
+	{
+		line = malloc(len);
+		rewind(fd);
+		TEST_ASSERT_EQUAL_INT(6, getline(&line, &len, fd));
+		TEST_ASSERT_EQUAL_INT(50, len);
+		TEST_ASSERT_EQUAL_STRING("line1\n", line);
+		free(line);
+	}
+	fclose(fd);
+}
+
+
+TEST(line, getline_longline)
+{
+	char *line = NULL;
+	size_t len = 50; /* allocated memory exceeds one demanded for a line that to be read */
```

Comment:
Allocated? There's no malloc.

Reply:
Yes. My copypaste error. 

Here the memory should be allocated by `getline()` by passing it arbitrary zero as length. But passing to low value also forces it to reallocate memory, even if it is not there in the first place. 

Changed len to `0`